pkg/idxd/plugin.go

```diff
@@ -0,0 +1,202 @@
+)
+
+const (
+	// Character devices directory.
+	charDevDir = "/dev/char"
+	// Frequency of device scans.
+	scanFrequency = 5 * time.Second
+)
+
+// getDevNodesFunc type allows overriding filesystem APIs (os.Stat, stat.Sys, etc) in tests.
+type getDevNodesFunc func(devDir, charDevDir, wqName string) ([]pluginapi.DeviceSpec, error)
+
+// DevicePlugin defines properties of the idxd device plugin.
+type DevicePlugin struct {
+	sysfsDir     string
+	statePattern string
+	devDir       string
+	charDevDir   string
+	sharedDevNum int
+	scanTicker   *time.Ticker
+	scanDone     chan bool
+	getDevNodes  getDevNodesFunc
+}
+
+// NewDevicePlugin creates DevicePlugin.
+func NewDevicePlugin(sysfsDir, statePattern, devDir string, sharedDevNum int) *DevicePlugin {
+	return &DevicePlugin{
+		sysfsDir:     sysfsDir,
+		statePattern: statePattern,
+		devDir:       devDir,
+		charDevDir:   charDevDir,
+		sharedDevNum: sharedDevNum,
+		scanTicker:   time.NewTicker(scanFrequency),
+		scanDone:     make(chan bool, 1),
+		getDevNodes:  getDevNodes,
+	}
+}
+
+// Scan discovers devices and reports them to the upper level API.
+func (dp *DevicePlugin) Scan(notifier dpapi.Notifier) error {
+	defer dp.scanTicker.Stop()
+	for {
+		devTree, err := dp.scan()
+		if err != nil {
+			return err
+		}
+
+		notifier.Notify(devTree)
+
+		select {
+		case <-dp.scanDone:
+			return nil
+		case <-dp.scanTicker.C:
+		}
+	}
+}
+
+func readFile(fpath string) (string, error) {
+	data, err := ioutil.ReadFile(fpath)
+	if err != nil {
+		return "", errors.WithStack(err)
+	}
+	return strings.TrimSpace(string(data)), nil
+}
+
+// getDevNodes collects device nodes that belong to working queue.
+func getDevNodes(devDir, charDevDir, wqName string) ([]pluginapi.DeviceSpec, error) {
+	// check if /dev/dsa/<work queue> device node exists
+	devPath := path.Join(devDir, wqName)
+	stat, err := os.Stat(devPath)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	// Check if it's a character device
+	if stat.Mode()&os.ModeCharDevice == 0 {
+		return nil, errors.Errorf("%s is not a character device", devPath)
+	}
+
+	// get /dev/char/<major>:<minor> symlink for the device node
+	// as libaccel-config requires it
+	rdev := stat.Sys().(*syscall.Stat_t).Rdev
+	charDevPath := path.Join(charDevDir, fmt.Sprintf("%d:%d", unix.Major(rdev), unix.Minor(rdev)))
+	stat, err = os.Lstat(charDevPath)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	if stat.Mode()&os.ModeSymlink == 0 {
+		return nil, errors.Errorf("%s is not a symlink", charDevPath)
+	}
+	// Check if symlink points to the correct device node
+	destPath, err := filepath.EvalSymlinks(charDevPath)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	if destPath != devPath {
+		return nil, errors.Errorf("%s points to %s instead of device node %s", charDevPath, destPath, devPath)
+	}
+
+	// report device node and /dev/char/<major>:<minor> symlink
+	// as libaccel-config works with a symlink
```

Comment:
the symlink doesn't need to be a device mount, does it?

Reply:
It does as far as I remember as libaccel-config works with it. I can re-check.

Reply:
OK. with SGX I was able to build `/dev/sgx/enclave -> /dev/sgx_enclave` backwards compatibility with just the latter being a device mount

Reply:
Ah, now I understand. I think it would be possible to do the same for DSA. I'm not sure it will be simpler. I can try.

Reply:
And to clarify what I meant to be sure:

```
[]pluginapi.DeviceSpec{
		{
			HostPath:      devPath,
			ContainerPath: devPath,
			Permissions:   "rw",
		},
}
and
[]pluginapi.Mount{
		{
			HostPath:      charDevPath,
			ContainerPath: charDevPath,
			Permissions:   "rw",
		},
}

Reply:
@mythi we can get rid of /dev/char mount if we remove this code:
```
	// Check if symlink points to the correct device node
	destPath, err := filepath.EvalSymlinks(charDevPath)
	if err != nil {
		return nil, errors.WithStack(err)
	}
	if destPath != devPath {
		return nil, errors.Errorf("%s points to %s instead of device node %s", charDevPath, destPath, devPath)
	}
```
i.e. if we don't check if symlink exists and points to the correct device node.

My idea was to validate symlink before sending its path to Kubelet.
wdyt?

Reply:
> My idea was to validate symlink before sending its path to Kubelet.

the validation is good to have. I'm not suggesting to get rid of the mount but change the symlink mount to `[]pluginapi.Mount` (instead of `[]pluginapi.DeviceSpec`)

Reply:
ah, ok. will try. thanks for explanations.

Reply:
agreed to fix this later

Reply:
Docker fails to undestand mounts with ':' inside file name: ```
Error response from daemon: invalid volume specification: '/dev/char/235:0:/dev/char/235:0'```

So we can't have /dev/char/<major>:<minor> as a mapping easily. It would require to fix it on a kubernetes level (dockershim or Kubelet I guess).